src/gleam/time/timestamp.gleam

```diff
@@ -227,12 +265,388 @@ fn to_civil(minutes: Int) -> #(Int, Int, Int) {
+    <<".", byte, remaining_bytes:bytes>>
+      if byte_zero <= byte && byte <= byte_nine
+    -> {
+      do_parse_second_fraction_as_nanoseconds(
+        from: <<byte, remaining_bytes:bits>>,
+        acc: 0,
+        pow: nanoseconds_per_second,
+        k: 0,
+      )
+    }
+    // bytes starts with a ".", which should introduce a fraction, but it does
+    // not, and so it is an ill-formed input.
+    <<".", _:bytes>> -> Error(Nil)
+    // bytes does not start with a "." so there is no fraction.  Call this 0
+    // nanoseconds.
+    _ -> Ok(#(0, bytes))
+  }
+}
+
+fn do_parse_second_fraction_as_nanoseconds(from bytes, acc acc, pow pow, k k) {
+  case bytes {
+    <<byte, remaining_bytes:bytes>> if byte_zero <= byte && byte <= byte_nine -> {
+      // Each digit place to the left in the fractional second is 10x fewer
+      // nanoseconds.
+      let pow = pow / 10
+
+      case int.compare(pow, 1) {
+        order.Lt -> {
+          // We already have the max precision for nanoseconds. Truncate any
+          // remaining digits.
+          do_parse_second_fraction_as_nanoseconds(
+            from: remaining_bytes,
+            acc:,
+            pow:,
+            k: k + 1,
+          )
+        }
+        order.Gt | order.Eq -> {
+          let digit = byte - 0x30
+          do_parse_second_fraction_as_nanoseconds(
+            from: remaining_bytes,
+            acc: acc + digit * pow,
+            pow:,
+            k: k + 1,
+          )
+        }
+      }
+    }
+    _ -> Ok(#(acc, bytes))
+  }
+}
+
+fn parse_offset(from bytes: BitArray) -> Result(#(Int, BitArray), Nil) {
+  case bytes {
+    <<"Z", remaining_bytes:bytes>> | <<"z", remaining_bytes:bytes>> ->
+      Ok(#(0, remaining_bytes))
+    _ -> parse_numeric_offset(bytes)
+  }
+}
+
+fn parse_numeric_offset(from bytes: BitArray) -> Result(#(Int, BitArray), Nil) {
+  use #(sign, bytes) <- result.try(parse_sign(from: bytes))
+  use #(hours, bytes) <- result.try(parse_hours(from: bytes))
+  use bytes <- result.try(accept_byte(from: bytes, value: byte_colon))
+  use #(minutes, bytes) <- result.try(parse_minutes(from: bytes))
+
+  let offset_seconds = offset_to_seconds(sign, hours:, minutes:)
+
+  Ok(#(offset_seconds, bytes))
+}
+
+fn parse_sign(from bytes) {
+  case bytes {
+    <<"+", remaining_bytes:bytes>> -> Ok(#("+", remaining_bytes))
+    <<"-", remaining_bytes:bytes>> -> Ok(#("-", remaining_bytes))
+    _ -> Error(Nil)
+  }
+}
+
+fn offset_to_seconds(sign, hours hours, minutes minutes) {
+  let abs_seconds = hours * seconds_per_hour + minutes * seconds_per_minute
+
+  case sign {
+    "-" -> -abs_seconds
+    _ -> abs_seconds
+  }
+}
+
+/// Parse and return the given number of digits from the given bytes.
+/// 
+fn parse_digits(
+  from bytes: BitArray,
+  count count: Int,
+) -> Result(#(Int, BitArray), Nil) {
+  do_parse_digits(from: bytes, count:, acc: 0, k: 0)
+}
+
+fn do_parse_digits(
+  from bytes: BitArray,
+  count count: Int,
+  acc acc: Int,
+  k k: Int,
```

Comment:
What's k?

Reply:
k is just the name of the loop counter. I could change the name to something more descriptive?